### PR TITLE
refactor: middleware

### PR DESCRIPTION
**Pull request by RealShadowNova**
This PR refactors how middleware's context is set and used, it also adds more functionality with decorators.

# Changes

- Remove `Josh.Options#middlewareContextData`
- Add another parameter to `Middleware`'s constructor
- Add `Josh.Options#autoEnsure` property which is the `AutoEnsure.ContextData#defaultValue` and will add `AutoEnsure` as a middleware
- Rename `CoreAutoEnsure` to `AutoEnsure`
- Add a warning if a non-middleware is provided in `Josh.Options#middlewares`
- Delete `BuiltInMiddleware` as it wasn't needed
- Add `reflect-metadata`, package is self-explanatory
- Add `@PreProvider` and `@PostProvider` decorators which adds another way to tell middleware when methods are ran
- Remove `Middleware.ContextData`
- Add generic `ContextData` to `Middleware` so that we don't need to declare `Middleware#context`

# Example Usage

## AutoEnsure
```javascript
new Josh({
  name: 'name',
  autoEnsure: { defaultValue: 'defaultValue' }
});
```

## Custom Middleware class

```typescript
@ApplyMiddlewareOptions({ name: 'my-middleware' })
export class MyMiddleware<StoredValue = unknown> extends Middleware<NonNullObject, StoredValue> {
  @PreProvider()
  public [Method.Set]<Value = StoredValue>(payload: Payloads.Set<Value>): Payloads.Set<Value> {
    // Runs when 'payload.trigger' is 'Trigger.PreProvider'
  }

  @PostProvider()
  public [Method.Get]<Value = StoredValue>(payload: Payloads.Get<Value>): Payloads.Get<Value> {
    // Runs when 'payload.trigger' is 'Trigger.PostProvider'
  }

  @PreProvider()
  @PostProvider()
  public [Method.Values](payload: Payloads.Values): Payloads.Values {
    // Runs when 'payload.trigger' is either 'Trigger.PreProvider' or 'Trigger.PostProvider'
  }
}
```

## Using Middleware

```javascript
new Josh({
  name: 'name',
  middlewares: [new MyMiddleware()]
});
```

```javascript
josh.use(new MyMiddleware());
```

### CODE REVIEW

File: src/lib/structures/Josh.ts [hunk 1]
```diff
@@ -79,6 +78,9 @@ export class Josh<StoredValue = unknown> {
       );
 
     this.middlewares = new MiddlewareStore({ instance: this });
+
+    if (middlewares !== undefined)
+      for (const middleware of middlewares.filter((middleware) => middleware instanceof Middleware)) this.use(middleware);
```
Comment (dan-online) [inline src/lib/structures/Josh.ts:83]: Perhaps throw an error if an inputted "middleware" is not an instance of Middleware, or at least a warning

File: src/lib/structures/Middleware.ts [hunk 2]
```diff
@@ -215,6 +217,18 @@ export class Middleware<StoredValue = unknown> {
     return payload;
   }
 
+  /**
+   * Sets the context for this middleware.
```
Comment (dan-online) [inline src/lib/structures/Middleware.ts:221]: More description for ``context`` needed

File: src/lib/structures/Josh.ts [hunk 3]
```diff
@@ -123,6 +125,11 @@ export class Josh<StoredValue = unknown> {
    * ```javascript
    * josh.use(new MyMiddleware());
    * ```
+   *
+   * @example
+   * ```javascript
+   * josh.use(new MyMiddleware().setContext({ data: 'data' }));
```
Comment (dan-online) [inline src/lib/structures/Josh.ts:131]: Move setting of context to the constructor perhaps (discussed in dsc)

Comment (dan-online) [reply src/lib/structures/Josh.ts:131]: "Josh.Options#autoEnsure, if they have the option there it adds auto ensure and so on for other built-in middleware" - @RealShadowNova

**Review by dan-online — CHANGES_REQUESTED** [review]
Just a few things and the context/setContext changes

File: src/middlewares/AutoEnsure.ts [hunk 4]
```diff
@@ -1,84 +1,81 @@
-import { ApplyMiddlewareOptions } from '../lib/decorators/ApplyMiddlewareOptions';
+import { ApplyMiddlewareOptions } from '../lib/decorators';
 import { Middleware } from '../lib/structures/Middleware';
-import { BuiltInMiddleware, Method, Payloads } from '../lib/types';
+import { Method, Payloads } from '../lib/types';
 
 @ApplyMiddlewareOptions({
-  name: BuiltInMiddleware.AutoEnsure,
+  name: 'autoEnsure',
   position: 0,
```
Comment (dan-online) [inline src/middlewares/AutoEnsure.ts:7]: What does position do?

**Review by dan-online — APPROVED** [review]
Just a couple questions but everything else looks good

File: src/middlewares/AutoEnsure.ts (hunk 4 above)
Comment (RealShadowNova) [reply src/middlewares/AutoEnsure.ts:7]: Tells Josh what order to run middleware in

File: src/lib/structures/Josh.ts [hunk 5]
```diff
@@ -73,6 +74,18 @@ export class Josh<StoredValue = unknown> {
     if (!(this.provider instanceof JoshProvider)) emitWarning(this.error(Josh.Identifiers.InvalidProvider));
 
     this.middlewares = new MiddlewareStore({ instance: this });
+
+    if (autoEnsure !== undefined) this.use(new AutoEnsure<StoredValue>(autoEnsure));
+    if (middlewares !== undefined)
```
Comment (dan-online) [inline src/lib/structures/Josh.ts:79]: Can we also check Array.isArray(middlewares) or something similar

File: src/lib/structures/Josh.ts [hunk 6]
```diff
@@ -73,6 +74,18 @@ export class Josh<StoredValue = unknown> {
...
+      for (const middleware of middlewares.filter((middleware) => {
+        if (!(middleware instanceof Middleware))
+          emitWarning(
+            this.error({ identifier: Josh.Identifiers.InvalidMiddleware, message: 'The middleware must extend the exported "Middleware" class.' })
```
Comment (dan-online) [inline src/lib/structures/Josh.ts]: Needs to be changed due to #57

File: src/lib/structures/Middleware.ts [hunk 7]
```diff
@@ -58,12 +58,18 @@ export class Middleware<StoredValue = unknown> {
...
 
-  public constructor(options: Middleware.Options) {
+  protected context: ContextData;
+
+  public constructor(context: ContextData, options: Partial<Middleware.Options> = {}) {
+    this.context = context;
+
     const { name, position, conditions } = options;
 
+    if (!name) throw new JoshError({ identifier: Middleware.Identifiers.NameNotFound, message: 'No name was provided for this middleware.' });
```
Comment (dan-online) [inline src/lib/structures/Middleware.ts:68]: Needs to be changed due to #57

**Review by dan-online — CHANGES_REQUESTED** [review]

File: src/lib/structures/Middleware.ts (hunk 7 above)
Comment (RealShadowNova) [reply src/lib/structures/Middleware.ts:68]: There is no `error()` method on `Middleware`

Comment (dan-online) [reply src/lib/structures/Middleware.ts:68]: Ah yep I see, coolio

**Review by dan-online — APPROVED** [review]
LGTM

**Pull request merged**